app/actions/remote/preference.ts

```diff
@@ -149,3 +151,34 @@ export const deleteSavedPost = async (serverUrl: string, postId: string) => {
         return {error};
     }
 };
+
+export const makeDirectChannelVisible = async (serverUrl: string, channelId: string, visible = true) => {
+    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
+    if (!database) {
+        return {error: `${serverUrl} database not found`};
+    }
+
+    try {
+        const channel = await getChannelById(database, channelId);
+        if (channel?.type === General.DM_CHANNEL || channel?.type === General.GM_CHANNEL) {
+            const userId = await getCurrentUserId(database);
+            const category = channel.type === General.DM_CHANNEL ? CATEGORY_DIRECT_CHANNEL_SHOW : CATEGORY_GROUP_CHANNEL_SHOW;
+            const name = channel.type === General.DM_CHANNEL ? getUserIdFromChannelName(userId, channel.name) : channelId;
+            const records = await queryPreferencesByCategoryAndName(database, category, name, (!visible).toString()).fetch();
+            if (records.length) {
+                const pref: PreferenceType = {
+                    user_id: userId,
+                    category,
+                    name,
+                    value: visible.toString(),
+                };
+                return savePreference(serverUrl, [pref]);
+            }
+        }
```

Comment:
this is writing to the database. not sure what you mean

Reply:
The way I read this is that we query the preference of the channel with the visibility contrary to what we are going to set. If we find any preference like that, we write to the database. If not, we assume the visibility is already set to what we are looking for, so we don't write.

So the undefined state is not properly handled here, and we are doing one query to check if we should write, instead of directly write.